[tools/codesign][tvOS][Jailbreak] Fake sign kodi-topshelf + fix crash on start on Unc0verTV

sy6sy2 opened the pull request:
## Description
<!--- Provide a general summary of your change in the Pull Request title above -->
<!--- Describe your change in detail here -->
This PR is related to this one: https://github.com/xbmc/xbmc/pull/18315 (more precisely is a missing part of this one...)
I noticed that the top-shelf binary also need to be fake signed in order to work on jailbroken Apple TV.

Also, Kodi is completely broken on Unc0verTV jailbroken devices, because we need to fake-sign ALL binaries (including frameworks, dylib, ...)

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here -->
* This PR allow to have TopShelf working on jailbroken Apple TV devices (when you install Kodi with NitoTV or with `dpkg -i`).
* This PR allow Unc0verTV devices to run Kodi

## How Has This Been Tested?
<!--- Please describe in detail how you tested your change -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc -->
On my jailbroken Apple TV 4K.
* Signed sideload --> Ok
* Fake-signed `dpkg -i` --> Ok

Also tested by @BTopbas on an Unc0verTV Apple TV 4K device.

## User impact

* tvOS jailbroken Kodi users will see Kodi Topshelf feature if they put Kodi app in the first line of their Apple TV home screen.
* Kodi will works on Unc0verTV devices.

## Screenshots (if appropriate):
![84372190-09deb800-abdb-11ea-8ef3-07117737442a](https://user-images.githubusercontent.com/11562279/91745669-a3ef4480-ebbb-11ea-9be1-8a631d1bb505.png)


## Types of change
<!--- What type of change does your code introduce? Put an `x` in all the boxes that apply like this: [X] -->
- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [ ] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
<!--- Go over all the following points, and put an `X` in all the boxes that apply like this: [X] -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [ ] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [X] All new and existing tests passed

fuzzard reviewed (commented):
Do we need to do an if [ -f .../kodi-topshelf ], or encase it  in a tvos platform check? Does ldid error if file doesn't exist?

Other comment would be around hardcoding Kodi in the filename. I can't remember if you can change topshelf app name, but I assume you can

sy6sy2 commented:
You are right, I will add the `if` because there is no topshelf file for iOS.
ldid return 1 code when file doesn't exist.

fuzzard commented:
Looks good to me

sy6sy2 committed:
[tools/codesign][darwin_embedded] Remove obsolete Kodi.xcent file

This file is no more generated by Xcode since version 9.3

sy6sy2 committed:
[tools/codesign][tvOS] Fake sign kodi-topshelf and other binaries

Fake-signing kodi-topshelf binary is needed to have
Kodi Top Shelf feature working on jailbroken devices.
Fake-signing all other Mach-O files is needed to have
Kodi working on Unc0verTV jailbroken devices.
(But it's not needed on checkra1n devices...)

fuzzard commented on tools/darwin/Support/Codesign.command:
```diff
@@ -103,4 +98,21 @@ if ! [ -z "${CODE_SIGN_IDENTITY_FOR_ITEMS}" ]; then
+
+  # On tvOS we also need to fake sign the Top Shelf binary
+  if [ "$APPLETVOS" ]; then
+    if [ -d "${TOPSHELF_PATH}/_CodeSignature" ]; then
```
Is this actually needed at this point?

fuzzard commented on tools/darwin/Support/Codesign.command:
```diff
@@ -103,4 +98,21 @@ if ! [ -z "${CODE_SIGN_IDENTITY_FOR_ITEMS}" ]; then
+    if [ -d "${TOPSHELF_PATH}/_CodeSignature" ]; then
+      rm -r "${TOPSHELF_PATH}/_CodeSignature"
+    fi
+    if [ -f "${TOPSHELF_PATH}/embedded.mobileprovision" ]; then
```
Is this actually needed? If it falls through to this codepath, it's already been deleted hasn't it?

sy6sy2 replied:
Do you mean that `_CodeSignature` and `embedded.mobileprovision` will always not exist in case `CODE_SIGN_IDENTITY_FOR_ITEMS` is empty?

fuzzard replied:
They are deleted at line 43 and 49 respectively. As it doesn't run codesign again in this script unless proper signing is provided, they won't be regenerated

sy6sy2 replied:
Line 43 and 49 delete the Kodi one but not the TopShelf one, no?

fuzzard replied:
Sorry, my bad. If that's the case, do a find and rm the result. Will do it in one hit. Doesn't need to be behind a platform check.
Does ldid actually care if it's there though?

sy6sy2 replied:
Like that --> last push?

ldid doesn't care, but on jailbroken devices those folder/files are not needed ; I just rm them to have a "clean" deb file without any signing stuff. Then the end user can directly see that he can (i) directly install it on jailbroken device or (ii) sign it with iOS App Signed.

fuzzard replied:
The only reason we delete them is for Dev use. A normal user (or a Jenkins build) will never have them anyway

sy6sy2 replied:
Oh yes, I always forget that Jenkins do not perform signing stuff, so you are right, those files should never be deb file produced by Jenkins.
So, to conclude, should we remove all the lines that delete `_CodeSignature` and `embedded.mobileprovision`?

I simply delete the "if appletvos" line at the end of the file.

sy6sy2 replied:
I moved the "delete" lines inside the "if ${CODE_SIGN_IDENTITY_FOR_ITEMS}", so that we don't care about `_CodeSignature` and `embedded.mobileprovision` files when we do not have signing stuff like Jenkins.

Also, I noticed a `libass` binary in Frameworks folder. ATM this one is not fake-signed and have "not signed at all". Not sure if it works on jailbroken devices ; how can I test this binary during Kodi runtime?

fuzzard reviewed (approved):
Happy with that, looks good

sy6sy2 commented:
Cool 😉  Thank you for your help.
@kambala-decapitator, is it ok for you?

sy6sy2 commented:
Jenkins build this please

kambala-decapitator commented on tools/darwin/Support/Codesign.command:
```diff
@@ -2,66 +2,49 @@
-GEN_ENTITLEMENTS="$NATIVEPREFIX/bin/gen_entitlements.py"
 DARWIN_EMBEDDED_ENTITLEMENTS="$XBMC_DEPENDS/share/darwin_embedded_entitlements.xml"
 LDID="$NATIVEPREFIX/bin/ldid"
+TOPSHELF_BINARY="${BUILT_PRODUCTS_DIR}/${EXECUTABLE_FOLDER_PATH}/PlugIns/${TOPSHELF_EXTENSION_NAME}.appex/${TOPSHELF_EXTENSION_NAME}"
```
1. I suggest to use `CONTENTS_PATH` defined below
2. correct way to find the right exexcutable is to read `CFBundleExecutable` from Info.plist

kambala-decapitator commented on tools/darwin/Support/Codesign.command:
```diff
@@ -2,66 +2,49 @@
-
-if [ ! "$MACOS" ]; then
-  ${GEN_ENTITLEMENTS} "${BUNDLEID}" "${BUILT_PRODUCTS_DIR}/${EXECUTABLE_FOLDER_PATH}/${EXECUTABLE_NAME}.xcent"
-  if [ -f "${CONTENTS_PATH}/embedded.mobileprovision" ]; then
-    rm -f "${CONTENTS_PATH}/embedded.mobileprovision"
-  fi
```
not sure why you want to delete this

sy6sy2 replied:
Because I use the find command

kambala-decapitator replied:
no need to use if, just use `rm -rf`

kambala-decapitator commented on tools/darwin/Support/Codesign.command:
```diff
@@ -2,66 +2,49 @@
-fi
+echo "Code sign identity is '${CODE_SIGN_IDENTITY_FOR_ITEMS}'"
 
-# delete existing codesigning
-if [ -d "${CONTENTS_PATH}/_CodeSignature" ]; then
-  rm -r "${CONTENTS_PATH}/_CodeSignature"
-fi
```
not sure why you want to delete this

sy6sy2 replied:
Because I use the find command

kambala-decapitator replied:
no need to use if, just use `rm -rf`

kambala-decapitator commented on tools/darwin/Support/Codesign.command:
```diff
@@ -2,66 +2,49 @@
-    echo "Doing a full bundle sign using genuine identity ${CODE_SIGN_IDENTITY_FOR_ITEMS}"
+    echo "Doing a full bundle sign using genuine identity '${CODE_SIGN_IDENTITY_FOR_ITEMS}'"
+    
+    # Delete existing codesign and provisionong file
+    # (only with maxdepth=1 because we want to keep existing TopShelf binary signature)
+    find "${CONTENTS_PATH}" -maxdepth 1 -name "embedded.mobileprovision" -exec rm {} \;
+    find "${CONTENTS_PATH}" -maxdepth 1 -name "_CodeSignature" -exec rm -r {} \;
```
I don't find this correct. These files should be deleted before deciding if we're signing or not.

Besides, the files reside in predefined locations, no need to use `find`.

sy6sy2 replied:
Why this is not correct? We care about these files only when we perform real signing, no? If we perform fake signing we don't care if these files are here or not?

I use `find` to handle the case where theses files do not exist.

kambala-decapitator replied:
Imagine the following use case. You built a signed app, then changed project configuration to build a non-signed app. You already have app bundle with code signature: with your change existing code signature won't be deleted which is wrong.

Using `-f` flag on `rm` won't spit error message and return error code. No need to use `find`, it's not meant for that.

sy6sy2 replied:
Ok I understand what you said. For you it's wrong to packages  _CodeSignature and embedded.mobileprovision in a deb file that will be fake signed. I'm right?

I will replace find by rm -f and move these lines out of the real signing "if" :-)

kambala-decapitator replied:
absolutely correct!

kambala-decapitator commented on tools/darwin/Support/Codesign.command:
```diff
@@ -103,4 +86,15 @@ if ! [ -z "${CODE_SIGN_IDENTITY_FOR_ITEMS}" ]; then
+  # See http://www.saurik.com/id/8
+  echo "Doing a fake sign using ldid for jailbroken devices"
+
+  "${LDID}" -S"${DARWIN_EMBEDDED_ENTITLEMENTS}" "${BUILT_PRODUCTS_DIR}/${EXECUTABLE_FOLDER_PATH}/${EXECUTABLE_NAME}"
```
use `CONTENTS_PATH`

sy6sy2 commented:
Hi, I finally create a "PostBuild" script for TopShelf in order to fake sign it if needed. I find this approach cleaner because we do not pollute main Codesign file with tvOS topShelf specific stuff.

Also, I added some code in order to fake-sign all Mach-O files. This step fix an issue where Unc0verTV users have Kodi crashing on start.
ATM Top-Shelf still not working on Unc0verTV devices, but I'm currently in contact with JoshTV trying to resolve this issue. Also, a big thanks to @BTopbas for all the tests he did because he is the only one here with a Unc0verTV jailbroken devices.

Sorry for the code changes...

kambala-decapitator commented on tools/darwin/Support/Codesign.command:
```diff
@@ -2,66 +2,47 @@
-  rm -r "${CONTENTS_PATH}/_CodeSignature"
-fi
+# Delete existing codesign and provisioning file
+rm -f "${CODESIGNING_FOLDER_PATH}/embedded.mobileprovision"
+rm -rf "${CODESIGNING_FOLDER_PATH}/_CodeSignature"
```
```suggestion
rm -f "${CONTENTS_PATH}/embedded.mobileprovision"
rm -rf "${CONTENTS_PATH}/_CodeSignature"
```

kambala-decapitator commented on tools/darwin/Support/Codesign.command:
```diff
@@ -103,4 +84,18 @@ if ! [ -z "${CODE_SIGN_IDENTITY_FOR_ITEMS}" ]; then
+
+  # All Mach-O files (except TopShelf)
+  for f in $(find "${CONTENTS_PATH}/AppData" "${CONTENTS_PATH}/Frameworks" -type f); do
+    if [[ $(file $f) == *"Mach-O"* ]]; then
+      "${LDID}" -S "${f}"
```
I suggest to stick to a single code style whenever possible: use either `$f` or `${f}` in both places

fuzzard commented on tools/darwin/Support/Codesign-topshelf.command:
```diff
@@ -0,0 +1,19 @@
+  echo "Doing a fake sign of Top Shelf binary using ldid for jailbroken devices"
+  "${LDID}" -S "${CODESIGNING_FOLDER_PATH}/${EXECUTABLE_NAME}"
+else
+  echo "Doing a real sign of Top Shelf binary (Xcode will do after)"
```
Dont bother with the else logging. Just adds extra output for no real use imo

sy6sy2 replied:
I will delete it.

fuzzard commented on tools/darwin/Support/Codesign-topshelf.command:
```diff
@@ -0,0 +1,19 @@
+LDID="$NATIVEPREFIX/bin/ldid"
+
+# Delete existing codesign and provisioning file
+rm -f "${CODESIGNING_FOLDER_PATH}/embedded.mobileprovision"
+rm -rf "${CODESIGNING_FOLDER_PATH}/_CodeSignature"
```
Im not sure this should be done in a regular signing environment. Is there a need for this outside of jailbreak usage? I havent come across the same signing issues that forces the need for this in the main app bundle, as xcode manages it all top to bottom for the topshelf appex.

Essentially, should this be inside the fake sign if check below?

fuzzard replied:
Also, does CODESIGNING_FOLDER_PATH actually give you the path to the topshelf appex in this context? Or is it only giving you the path to the main app bundle?

sy6sy2 replied:
If I understand correctly your question, this conversation from @kambala-decapitator should answer your question? https://github.com/xbmc/xbmc/pull/18335#discussion_r482011235

sy6sy2 replied:
Concerning `CODESIGNING_FOLDER_PATH`, this is what Jenkins gives:

```
+ rm -f /Users/Shared/jenkins/workspace/TVOS/build/build/Debug-appletvos/kodi-topshelf.appex/embedded.mobileprovision
+ rm -rf /Users/Shared/jenkins/workspace/TVOS/build/build/Debug-appletvos/kodi-topshelf.appex/_CodeSignature
```

So in TopShelf context `CODESIGNING_FOLDER_PATH  == /Users/Shared/jenkins/workspace/TVOS/build/build/Debug-appletvos/kodi-topshelf.appex/`

fuzzard replied:
All fine then, if it works

sy6sy2 commented:
Thank you both for your help with this PR.
I will merge it tomorrow if no objection 😉
